src/streams/openssl.c

```diff
@@ -40,14 +40,257 @@ SSL_CTX *git__ssl_ctx;
+
+char *(*_ERR_error_string)(unsigned long e, char *buf);
+void (*_ERR_error_string_n)(unsigned long e, char *buf, size_t len);
+unsigned long (*_ERR_get_error)(void);
+
+# define _sk_GENERAL_NAME_num(sk) _OPENSSL_sk_num((const OPENSSL_STACK *)sk)
+# define _sk_GENERAL_NAME_value(sk, idx) _OPENSSL_sk_value((const OPENSSL_STACK *)sk, idx)
+# define _GENERAL_NAMES_free(sk) _OPENSSL_sk_free((OPENSSL_STACK *)sk);
+
+int (*_OPENSSL_init_ssl)(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings);
+int (*_OPENSSL_sk_num)(const OPENSSL_STACK *sk);
+void *(*_OPENSSL_sk_value)(const OPENSSL_STACK *sk, int i);
+void (*_OPENSSL_sk_free)(OPENSSL_STACK *sk);
+
+# define _OPENSSL_malloc(num) _CRYPTO_malloc(num, OPENSSL_FILE, OPENSSL_LINE)
+# define _OPENSSL_free(addr) _CRYPTO_free(addr, OPENSSL_FILE, OPENSSL_LINE)
+
+int (*_SSL_connect)(SSL *ssl);
+long (*_SSL_ctrl)(SSL *ssl, int cmd, long arg, void *parg);
+void (*_SSL_free)(SSL *ssl);
+int (*_SSL_get_error)(SSL *ssl, int ret);
+X509 *(*_SSL_get_peer_certificate)(const SSL *ssl);
+long (*_SSL_get_verify_result)(const SSL *ssl);
+SSL *(*_SSL_new)(SSL_CTX *ctx);
+int (*_SSL_read)(SSL *ssl, const void *buf, int num);
+void (*_SSL_set_bio)(SSL *ssl, BIO *rbio, BIO *wbio);
+int (*_SSL_shutdown)(SSL *ssl);
+int (*_SSL_write)(SSL *ssl, const void *buf, int num);
+
+long (*_SSL_CTX_ctrl)(SSL_CTX *ctx, int cmd, long larg, void *parg);
+void (*_SSL_CTX_free)(SSL_CTX *ctx);
+SSL_CTX *(*_SSL_CTX_new)(const SSL_METHOD *method);
+int (*_SSL_CTX_set_cipher_list)(SSL_CTX *ctx, const char *str);
+int (*_SSL_CTX_set_default_verify_paths)(SSL_CTX *ctx);
+long (*_SSL_CTX_set_options)(SSL_CTX *ctx, long options);
+void (*_SSL_CTX_set_verify)(SSL_CTX *ctx, int mode, int (*verify_callback)(int, X509_STORE_CTX *));
+int (*_SSL_CTX_load_verify_locations)(SSL_CTX *ctx, const char *CAfile, const char *CApath);
+
+# define _SSL_CTX_set_mode(ctx, mode) _SSL_CTX_ctrl(ctx, SSL_CTRL_MODE, mode, NULL)
+
+# define _SSLv23_method _TLS_method
+
+const SSL_METHOD *(*_TLS_method)(void);
+
+ASN1_STRING *(*_X509_NAME_ENTRY_get_data)(const X509_NAME_ENTRY *ne);
+X509_NAME_ENTRY *(*_X509_NAME_get_entry)(X509_NAME *name, int loc);
+int (*_X509_NAME_get_index_by_NID)(X509_NAME *name, int nid, int lastpos);
+void (*_X509_free)(X509 *a);
+void *(*_X509_get_ext_d2i)(const X509 *x, int nid, int *crit, int *idx);
+X509_NAME *(*_X509_get_subject_name)(const X509 *x);
+
+int (*_i2d_X509)(X509 *a, unsigned char **ppout);
+
+GIT_INLINE(void *) openssl_sym(int *err, const char *name)
+{
+	void *symbol;
+
+	/* if we've seen an err, noop to retain it */
+	if (*err)
+		return NULL;
+
+	if ((symbol = dlsym(openssl_handle, name)) == NULL) {
+		const char *msg = dlerror();
+		git_error_set(GIT_ERROR_SSL, "could not load ssl function '%s': %s", name, msg ? msg : "unknown error");
+		*err = -1;
+	}
+
+	return symbol;
+}
+
+static int openssl_dynamic_init(void)
+{
+	int err = 0;
+
+	if ((openssl_handle = dlopen("libssl.so.1.1", RTLD_NOW)) == NULL) {
```

Comment:
I'm not sure this is flexible enough to work with all linux distros. For example, I recall that RHEL derivatives have a completely different naming scheme for their OpenSSL libraries. This [StackExchange](https://unix.stackexchange.com/a/421164) answer gives an example of what I'm talking about. The RHEL library is named `libssl.so.10`.

Would doing something like what is suggested in that answer be possible to dynamically look up the name?

Reply:
Yeah, we can definitely load `libssl.so.10`.  The tricky bit is going to be the API differences over the years, and identifying when to fallback to them because there's no "give me your version number" API in ye olde days.  I think that we can determine this based on identifying whether symbols exist or not, but it's not going to be particularly pretty.  😭

Reply:
I spent some more time looking into this, and it does seem that in the newest versions, RHEL/CentOS have decided to ship OpenSSL 1.1 with standard naming scheme, so they do have `libssl.so.1.1`. However, looking at [.NET 5 - Supported OS versions](https://github.com/dotnet/core/blob/main/release-notes/5.0/5.0-supported-os.md#linux), Microsoft is still supporting RHEL/CentOS 7, which just has the non-standard 1.0 name for the library. It doesn't look like this is changing for [.NET 6](https://github.com/dotnet/core/blob/main/release-notes/6.0/supported-os.md#linux) either.

Would it be possible to attempt to open `libssl.so.1.1` and if that fails try `libssl.so.10` and use that to determine which versions of the APIs to call?

Reply:
> Would it be possible to attempt to open libssl.so.1.1 and if that fails try libssl.so.10 and use that to determine which versions of the APIs to call?

Yes - this would be possible.  But it's not yet clear to me whether `libssl.so.10` always corresponds to a particular ABI - on my centos 7.5.1804 installation this is a link to `libssl.1.0.2k` but I don't know if that's a guarantee across all centos 7 minors and the corresponding rhel releases.  Need to do a little more investigation.